### PR TITLE
misc/mutex: fix assertion if nxmutex_reset() before nxmutex_unlock()

### PR DESCRIPTION
**Pull request by anchao**
## Summary

misc/mutex: fix assertion if nxmutex_reset() before nxmutex_unlock()

sim/rpserver

```
NuttShell (NSH) NuttX-12.0.0
server> cu
_assert: Current Version: NuttX server 12.0.0 3ead669e7a-dirty Feb  2 2023 23:53:48 sim
_assert: Assertion failed : at file: libs/libc/misc/lib_mutex.c:303 task: cu 0x5662fff4
```

## Impact

N/A

## Testing

sim/rpserver

### CODE REVIEW

**Comment by pkarashchenko** [comment]
We are "playing around" with nxmutex for a while. Maybe we can consider implementing a separate kernel object instead of relying on semaphores?

**Comment by anchao** [comment]
> We are "playing around" with nxmutex for a while. Maybe we can consider implementing a separate kernel object instead of relying on semaphores?

Haha, yes, separated kernel object is a correct route, we will consider and plan to refactor mutex in recently.

**Comment by xiaoxiang781216** [comment]
> We are "playing around" with nxmutex for a while. Maybe we can consider implementing a separate kernel object instead of relying on semaphores?

Unlock has to handle the abnormal case If mutex still support reset operation regardless whether you implement mutex from scratch or on top of semphare.

File: include/nuttx/mutex.h [hunk 1]
```diff
@@ -35,6 +35,7 @@
  ****************************************************************************/
 
 #define NXMUTEX_NO_HOLDER      ((pid_t)-1)
+#define NXMUTEX_RESET          ((pid_t)-2)
```
Comment (xiaoxiang781216) [inline include/nuttx/mutex.h:38]: move to source code

File: libs/libc/misc/lib_mutex.c [hunk 2]
```diff
@@ -329,15 +356,9 @@ int nxmutex_unlock(FAR mutex_t *mutex)
...
+  mutex->holder = NXMUTEX_RESET;
 
-  ret = nxsem_reset(&mutex->sem, 1);
-  if (ret >= 0)
-    {
-      mutex->holder = NXMUTEX_NO_HOLDER;
-    }
-
-  return ret;
+  return nxsem_reset(&mutex->sem, 1);
```
Comment (xiaoxiang781216) [inline libs/libc/misc/lib_mutex.c:361]: if we don't handle nxsem_reset error, it's better  change the return type of nxmutex_reset to void

File: libs/libc/misc/lib_mutex.c [hunk 3]
```diff
@@ -300,14 +322,19 @@ int nxmutex_unlock(FAR mutex_t *mutex)
...
+    }
+
   DEBUGASSERT(nxmutex_is_hold(mutex));
 
   mutex->holder = NXMUTEX_NO_HOLDER;
 
   ret = _SEM_POST(&mutex->sem);
   if (ret < 0)
     {
-      return _SEM_ERRVAL(ret);
```
Comment (xiaoxiang781216) [inline libs/libc/misc/lib_mutex.c:310]: should we restore the hold:
mutex->holder = _SCHED_GETTID();

File: include/nuttx/mutex.h (hunk 1 above)
Comment (anchao) [reply include/nuttx/mutex.h:38]: Done

File: libs/libc/misc/lib_mutex.c (hunk 2 above)
Comment (anchao) [reply libs/libc/misc/lib_mutex.c:361]: Done

File: libs/libc/misc/lib_mutex.c (hunk 3 above)
Comment (anchao) [reply libs/libc/misc/lib_mutex.c:310]: Done

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**